File: tools/parsing-question/src/parser/latex_parser.py
"""
Main LaTeX Question Parser

Combines all parsing components to extract complete question data from LaTeX content.
"""

import re
from typing import List, Optional, Tuple
import sys
import os

# Add parent directory to path for imports
current_dir = os.path.dirname(os.path.abspath(__file__))
parent_dir = os.path.dirname(current_dir)
sys.path.insert(0, parent_dir)

from models.question import Question
from models.question_code import QuestionCode
from parser.bracket_parser import BracketParser
from parser.content_extractor import ContentExtractor
from parser.answer_extractor import AnswerExtractor
from parser.question_code_parser import QuestionCodeParser


class LaTeXQuestionParser:
    """
    Main parser class that orchestrates the complete parsing process.
    
    Extracts questions from LaTeX content and converts them to structured Question objects.
    """
    
    @classmethod
    def extract_questions_from_content(cls, latex_content: str) -> List[str]:
        """
        Extract individual question blocks from LaTeX content.
        
        Args:
            latex_content: Full LaTeX file content
            
        Returns:
            List of question blocks (each containing one \\begin{ex}...\\end{ex})
        """
        # Find all \begin{ex}...\end{ex} blocks
        question_blocks = []
        
        # Use environment extraction from BracketParser
        ex_environments = BracketParser.extract_environment_content(latex_content, "ex")

        if not ex_environments:
            return []

        # Simple approach: find complete ex environments with regex
        pattern = r'((?:%.*?\n)*?)\\\\begin\\{ex\\}(.*?)\\\\end\\{ex\\}'
        matches = re.finditer(pattern, latex_content, re.DOTALL)

        for match in matches:
            metadata = match.group(1) if match.group(1) else ""
            ex_content = match.group(2)

            # Combine metadata + ex environment
            question_block = metadata + "\\\\begin{ex}" + ex_content + "\\\\end{ex}"
            question_blocks.append(question_block.strip())
        
        return question_blocks
    
    @classmethod
    def parse_single_question(cls, question_block: str) -> Optional[Question]:
        """
        Parse a single question block into a Question object.
        
        Args:
            question_block: LaTeX content for one question
            
        Returns:
            Question object or None if parsing fails
        """
        try:
            # Step 1: Extract metadata
            question_code, subcount, source = QuestionCodeParser.parse_metadata(question_block)
            
            # Step 2: Extract and clean content
            raw_content, clean_content = ContentExtractor.extract_clean_content(question_block)
            
            if not raw_content:
                return None
            
            # Step 3: Identify question type
            question_type = AnswerExtractor.identify_question_type(raw_content)
            
            # Skip MA (Matching) questions as requested
            if question_type == "MA":
                return None
            
            # Step 4: Extract answers and correctAnswer
            answers, correct_answer = AnswerExtractor.extract_answers_and_correct(raw_content, question_type)
            
            # Step 5: Extract solution
            solution = AnswerExtractor.extract_solution(raw_content)
            
            # Step 6: Create Question object
            question = Question(
                rawContent=raw_content.strip(),  # Chỉ lưu nội dung từ ex environment, không bao gồm source bên ngoài
                content=clean_content,
                type=question_type,
                subcount=subcount,
                source=source,
                answers=answers,
                correctAnswer=correct_answer,
                solution=solution,
                questionCodeId=question_code.code if question_code else None
            )
            
            return question
            
        except Exception as e:
            # Log error but don't crash the entire process
            print(f"Error parsing question: {str(e)}")
            return None
    
    @classmethod
    def parse_latex_file(cls, latex_content: str) -> Tuple[List[Question], List[QuestionCode], List[str]]:
        """
        Parse complete LaTeX file and extract all questions.
        
        Args:
            latex_content: Full LaTeX file content
            
        Returns:
            Tuple of (questions, question_codes, errors)
        """
        questions = []
        question_codes = []
        errors = []
        
        # Extract question blocks
        question_blocks = cls.extract_questions_from_content(latex_content)
        
        # Track unique question codes
        seen_codes = set()
        
        for i, block in enumerate(question_blocks):
            try:
                question = cls.parse_single_question(block)
                
                if question:
                    # Assign sequential ID
                    question.id = len(questions) + 1
                    questions.append(question)
                    
                    # Collect unique question codes
                    if question.questionCodeId and question.questionCodeId not in seen_codes:
                        question_code = QuestionCodeParser.extract_question_code(block)
                        if question_code:
                            question_codes.append(question_code)
                            seen_codes.add(question_code.code)
                
                else:
                    errors.append(f"Question block {i+1}: Failed to parse")
                    
            except Exception as e:
                error_msg = f"Question block {i+1}: {str(e)}"
                errors.append(error_msg)
                print(f"Error: {error_msg}")
        
        return questions, question_codes, errors
    
    @classmethod
    def validate_question(cls, question: Question) -> List[str]:
        """
        Validate a parsed question.
        
        Args:
            question: Question to validate
            
        Returns:
            List of validation errors
        """
        return question.validate()
    
    @classmethod
    def get_parsing_statistics(cls, questions: List[Question], errors: List[str]) -> dict:
        """
        Generate parsing statistics.
        
        Args:
            questions: List of parsed questions
            errors: List of parsing errors
            
        Returns:
            Dictionary with statistics
        """
        total_questions = len(questions) + len(errors)
        
        # Count by type
        type_counts = {}
        for question in questions:
            type_counts[question.type] = type_counts.get(question.type, 0) + 1
        
        # Count questions with answers
        questions_with_answers = sum(1 for q in questions if q.answers)
        questions_with_solutions = sum(1 for q in questions if q.solution)
        
        return {
            'total_blocks_processed': total_questions,
            'successful_parses': len(questions),
            'failed_parses': len(errors),
            'success_rate': len(questions) / total_questions * 100 if total_questions > 0 else 0,
            'question_types': type_counts,
            'questions_with_answers': questions_with_answers,
            'questions_with_solutions': questions_with_solutions,
            'unique_question_codes': len(set(q.questionCodeId for q in questions if q.questionCodeId))
        }

File: apps/backend/internal/latex/latex_parser.go
package latex

import (
	"fmt"
	"strings"

	"github.com/AnhPhan49/exam-bank-system/apps/backend/internal/entity"
	"github.com/google/uuid"
	"github.com/jackc/pgtype"
)

// LaTeXQuestionParser orchestrates the complete parsing process
type LaTeXQuestionParser struct {
	bp  *BracketParser
	qcp *QuestionCodeParser
	ce  *ContentExtractor
	ae  *AnswerExtractor
}

// NewLaTeXQuestionParser creates a new LaTeX question parser
func NewLaTeXQuestionParser() *LaTeXQuestionParser {
	return &LaTeXQuestionParser{
		bp:  NewBracketParser(),
		qcp: NewQuestionCodeParser(),
		ce:  NewContentExtractor(),
		ae:  NewAnswerExtractor(),
	}
}

// ParseLatexContent parses complete LaTeX file and extracts all questions
func (p *LaTeXQuestionParser) ParseLatexContent(latexContent string) ([]entity.Question, []entity.QuestionCode, []string) {
	var questions []entity.Question
	var questionCodes []entity.QuestionCode
	var errors []string

	// Extract question blocks
	questionBlocks := p.bp.ExtractEnvironmentContent(latexContent, "ex")

	// Track unique question codes
	seenCodes := make(map[string]bool)

	for i, block := range questionBlocks {
		question, questionCode, err := p.ParseSingleQuestion(block)
		if err != nil {
			errors = append(errors, fmt.Sprintf("Question block %d: %v", i+1, err))
			continue
		}

		if question != nil {
			// Assign sequential ID
			question.ID.Set(fmt.Sprintf("q_%d", len(questions)+1))
			questions = append(questions, *question)

			// Collect unique question codes
			if questionCode != nil && !seenCodes[questionCode.Code.String] {
				questionCodes = append(questionCodes, *questionCode)
				seenCodes[questionCode.Code.String] = true
			}
		}
	}

	return questions, questionCodes, errors
}

// ParseSingleQuestion parses a single question block into Question and QuestionCode entities
func (p *LaTeXQuestionParser) ParseSingleQuestion(questionBlock string) (*entity.Question, *entity.QuestionCode, error) {
	// Step 1: Extract metadata (questionCode, subcount, source)
	questionCode, subcount, source := p.qcp.ParseMetadata(questionBlock)

	// Step 2: Extract and clean content
	// Since questionBlock is already extracted from ex environment,
	// we need to apply content cleaning steps directly
	rawContent := strings.TrimSpace(questionBlock)
	if rawContent == "" {
		return nil, nil, fmt.Errorf("no content found in question block")
	}

	// Apply 7-step content cleaning process
	cleanContent := rawContent

	// Step 1: Remove \begin{ex} and \end{ex} tags (already done by BracketParser)
	// Step 2: Remove metadata patterns
	cleanContent = p.ce.RemoveMetadataPatterns(cleanContent)

	// Step 3: Handle layout commands
	cleanContent = p.ce.HandleLayoutCommands(cleanContent)

	// Step 4: Remove image commands
	cleanContent = p.ce.RemoveImageCommands(cleanContent)

	// Step 5: Remove answer commands and their content completely
	cleanContent = p.ce.RemoveAnswerCommands(cleanContent)

	// Step 6: Remove solution section
	cleanContent = p.ce.RemoveSolutionSection(cleanContent)

	// Step 7: Normalize whitespace
	cleanContent = p.ce.NormalizeWhitespace(cleanContent)

	// Validate clean content is not empty
	if strings.TrimSpace(cleanContent) == "" {
		return nil, nil, fmt.Errorf("no valid content found after cleaning")
	}

	// Step 3: Identify question type
	questionType := p.ae.IdentifyQuestionType(rawContent)

	// Validate question type
	if questionType == "" {
		return nil, nil, fmt.Errorf("unable to identify question type")
	}

	// Skip MA (Matching) questions as requested in design
	if questionType == string(entity.QuestionTypeMA) {
		return nil, nil, fmt.Errorf("matching questions (MA) are not supported yet")
	}

	// Step 4: Extract answers and correct answer
	answersJSON, correctAnswerJSON := p.ae.ExtractAnswersAndCorrect(rawContent, questionType)

	// Validate answers for MC and TF questions
	if (questionType == string(entity.QuestionTypeMC) || questionType == string(entity.QuestionTypeTF)) && len(answersJSON.Bytes) == 0 {
		return nil, nil, fmt.Errorf("no answers found for %s question", questionType)
	}

	// Step 5: Extract solution
	solution := p.ae.ExtractSolution(rawContent)

	// Step 6: Create Question entity
	question := &entity.Question{}

	// Set required fields
	question.RawContent.Set(rawContent)
	question.Content.Set(cleanContent)
	question.Type.Set(questionType)

	// Set optional fields
	if subcount != "" {
		question.Subcount.Set(subcount)
	} else {
		question.Subcount.Status = pgtype.Null
	}

	if source != "" {
		question.Source.Set(source)
	} else {
		question.Source.Status = pgtype.Null
	}

	if solution != "" {
		question.Solution.Set(solution)
	} else {
		question.Solution.Status = pgtype.Null
	}

	// Set answers and correct answer (JSONB fields)
	question.Answers = answersJSON
	question.CorrectAnswer = correctAnswerJSON

	// Set defaults
	question.UsageCount.Set(0)
	question.Creator.Set("ADMIN")
	question.Feedback.Set(0)
	question.Difficulty.Set(string(entity.QuestionDifficultyMedium))

	// Determine status based on content completeness
	question.Status.Set(p.determineQuestionStatus(questionType, answersJSON, correctAnswerJSON))

	// Set question code ID if available
	if questionCode != nil {
		question.QuestionCodeID.Set(questionCode.Code.String)
	} else {
		question.QuestionCodeID.Status = pgtype.Null
	}

	// Initialize empty tag array
	question.Tag.Set([]string{})

	return question, questionCode, nil
}

// ParseFromRawLatex parses raw LaTeX content and returns single question
func (p *LaTeXQuestionParser) ParseFromRawLatex(rawLatex string) (*entity.Question, error) {
	question, _, err := p.ParseSingleQuestion(rawLatex)
	if err != nil {
		return nil, err
	}

	if question == nil {
		return nil, fmt.Errorf("failed to parse question from LaTeX content")
	}

	// Generate UUID for ID
	question.ID.Set(uuid.New().String())

	return question, nil
}

// determineQuestionStatus determines question status based on completeness
func (p *LaTeXQuestionParser) determineQuestionStatus(questionType string, answers, correctAnswer pgtype.JSONB) string {
	switch questionType {
	case string(entity.QuestionTypeMC):
		if answers.Status == pgtype.Present && correctAnswer.Status == pgtype.Present {
			return string(entity.QuestionStatusActive)
		}
		return string(entity.QuestionStatusPending)

	case string(entity.QuestionTypeTF):
		if answers.Status == pgtype.Present {
			return string(entity.QuestionStatusActive)
		}
		return string(entity.QuestionStatusPending)

	case string(entity.QuestionTypeSA):
		if correctAnswer.Status == pgtype.Present {
			return string(entity.QuestionStatusActive)
		}
		return string(entity.QuestionStatusPending)

	case string(entity.QuestionTypeES):
		// ES questions are always active (no answers expected)
		return string(entity.QuestionStatusActive)

	default:
		return string(entity.QuestionStatusPending)
	}
}
